File: src/lib/truth/guard.ts
import type {
  ResumeFacts,
  TailoredResume,
  TruthReport,
  TruthViolation,
} from "@/lib/schemas";

/**
 * The truth guard runs after the model rewrites the resume and before anything
 * is rendered. It is deliberately deterministic: the model is asked to be
 * truthful, but the guard is what actually enforces it.
 *
 * Three classes of fabrication are what matter in practice:
 *   1. A line with no traceable origin in the uploaded resume.
 *   2. A metric that did not exist before the rewrite ("reduced latency 40%").
 *   3. A technology the candidate never claimed ("added Kubernetes").
 * Employer facts (company, title, dates) are checked for tampering too, since
 * those are the fields a recruiter verifies first.
 */

/** Technologies common enough that seeing one appear from nowhere is a red flag. */
const TECH_VOCABULARY = [
  // languages
  "javascript","typescript","python","java","go","golang","rust","ruby","php","c#","c++","kotlin","swift","scala","elixir","perl","r","dart","solidity",
  // frontend
  "react","angular","vue","svelte","next.js","nextjs","nuxt","remix","redux","zustand","tailwind","bootstrap","jquery","webpack","vite","astro",
  // backend
  "node.js","nodejs","express","nestjs","fastify","django","flask","fastapi","spring","spring boot","rails","laravel","gin","fiber","asp.net",
  // data
  "postgresql","postgres","mysql","mariadb","sqlite","mongodb","redis","elasticsearch","opensearch","cassandra","dynamodb","neo4j","clickhouse","snowflake","bigquery","redshift","supabase","firebase","prisma","typeorm","sequelize","drizzle","sqlalchemy","hibernate",
  // cloud / devops
  "aws","azure","gcp","google cloud","kubernetes","docker","terraform","ansible","jenkins","github actions","gitlab ci","circleci","argocd","helm","ec2","s3","lambda","ecs","eks","fargate","cloudfront","rds","sqs","sns","vercel","netlify","heroku","cloudflare","nginx","apache",
  // messaging / streaming
  "kafka","rabbitmq","celery","bullmq","sidekiq","nats","pulsar","airflow","temporal",
  // observability
  "prometheus","grafana","datadog","sentry","opentelemetry","new relic","splunk","kibana","jaeger",
  // api / protocol
  "graphql","grpc","rest","restful","websocket","websockets","soap","openapi","swagger","trpc","protobuf",
  // auth / security
  "oauth","oauth2","oidc","jwt","saml","keycloak","auth0","clerk","rbac","abac",
  // ai / ml
  "tensorflow","pytorch","keras","scikit-learn","sklearn","pandas","numpy","huggingface","langchain","llamaindex","openai","anthropic","claude","gpt","llm","rag","pinecone","weaviate","chromadb","qdrant","pgvector","mlflow","opencv","yolo","bert","transformers",
  // testing
  "jest","vitest","pytest","junit","cypress","playwright","selenium","mocha","chai","testing library","k6","jmeter",
  // other
  "git","github","gitlab","bitbucket","jira","linux","bash","graphite","stripe","razorpay","twilio","sendgrid","elastic","kong","istio","rabbit","socket.io",
];

/** Terms that mean the same thing to a human but differ as strings. */
const ALIASES: Record<string, string> = {
  golang: "go",
  postgres: "postgresql",
  nextjs: "next.js",
  nodejs: "node.js",
  sklearn: "scikit-learn",
  restful: "rest",
  websockets: "websocket",
  oauth2: "oauth",
  "google cloud": "gcp",
  "spring boot": "spring",
};

const canonical = (term: string) => {
  const t = term.toLowerCase().trim();
  return ALIASES[t] ?? t;
};

/** Lowercase, collapse punctuation that varies between writers, keep word shape. */
const normalize = (s: string) =>
  s
    .toLowerCase()
    .replace(/[‐-―]/g, "-")
    .replace(/[^a-z0-9+#./\- ]+/g, " ")
    .replace(/\s+/g, " ")
    .trim();

/**
 * Metric tokens: percentages, multipliers, durations, byte/req counts, money,
 * and bare numbers of two digits or more. Single digits are skipped because
 * "3 microservices" is usually a restatement, not an invented claim, and
 * flagging them buries the real findings in noise.
 */
const METRIC_PATTERN =
  /(\d[\d,.]*\s?(?:%|percent|x\b|ms\b|s\b|sec\b|seconds\b|min\b|minutes\b|hours?\b|days?\b|weeks?\b|months?\b|years?\b|k\b|m\b|b\b|mb\b|gb\b|tb\b|rps\b|qps\b|tps\b|req\/s|users?\b|customers?\b|clients?\b))|([$₹€£]\s?\d[\d,.]*)|(\b\d{2,}\b)/gi;

function extractMetrics(text: string): string[] {
  return (text.match(METRIC_PATTERN) ?? []).map((m) => m.trim().toLowerCase());
}

/** A metric matches its source if the same digits appear there in any form. */
function digitsOf(s: string) {
  return s.replace(/[^0-9]/g, "");
}

function buildFactIndex(facts: ResumeFacts) {
  const index = new Map<string, string>();
  const add = (id: string, text: string) => index.set(id, text);

  if (facts.summary) add("SUMMARY", facts.summary);
  if (facts.headline) add("HEADLINE", facts.headline);

  for (const e of facts.experience) {
    add(e.id, [e.company, e.title, e.location, e.startDate, e.endDate, ...e.tech].join(" "));
    for (const b of e.bullets) add(b.id, b.text);
  }
  for (const p of facts.projects) {
    add(p.id, [p.name, p.description, ...p.tech].join(" "));
    for (const b of p.bullets) add(b.id, b.text);
  }
  for (const ed of facts.education) {
    add(ed.id, [ed.institution, ed.degree, ed.dates, ed.details].join(" "));
  }
  for (const s of facts.skills) add(s.id, [s.category, ...s.items].join(" "));
  for (const c of facts.certifications) add(c.id, c.text);
  for (const o of facts.otherSections) {
    add(o.id, o.heading);
    for (const b of o.bullets) add(b.id, b.text);
  }
  return index;
}

/** Everything the candidate ever claimed, as one normalized haystack. */
function buildCorpus(facts: ResumeFacts, rawResumeText: string) {
  const index = buildFactIndex(facts);
  return normalize([...index.values()].join(" \n ") + " \n " + rawResumeText);
}

function techTermsIn(text: string, vocabulary: string[]): string[] {
  const hay = normalize(text);
  const found = new Set<string>();
  for (const term of vocabulary) {
    const t = normalize(term);
    if (t.length < 2) continue;
    // Word-ish boundary: the term must not be glued to surrounding letters.
    const re = new RegExp(`(^|[^a-z0-9])${t.replace(/[.*+?^${}()|[\]\\]/g, "\\$&")}([^a-z0-9]|$)`);
    if (re.test(hay)) found.add(canonical(term));
  }
  return [...found];
}

/**
 * The built-in list is a floor, not a ceiling — it cannot know about whatever
 * shipped last quarter. Seeding it with the job description's own vocabulary
 * means the terms this particular application will be judged on are always
 * covered, which is exactly where an invented claim would be most tempting.
 */
function buildVocabulary(jdTerms: string[]): string[] {
  const extra = jdTerms
    .map((t) => t.trim())
    .filter((t) => t.length >= 2 && t.length <= 40 && /[a-z]/i.test(t));
  return [...new Set([...TECH_VOCABULARY, ...extra])];
}

export function runTruthGuard(
  tailored: TailoredResume,
  facts: ResumeFacts,
  rawResumeText: string,
  jdTerms: string[] = [],
): TruthReport {
  const violations: TruthViolation[] = [];
  const factIndex = buildFactIndex(facts);
  const corpus = buildCorpus(facts, rawResumeText);
  const vocabulary = buildVocabulary(jdTerms);
  const corpusTech = new Set(techTermsIn(corpus, vocabulary));
  const corpusDigits = new Set(extractMetrics(corpus).map(digitsOf).filter(Boolean));

  const checkLine = (location: string, text: string, sourceIds: string[]) => {
    if (sourceIds.length === 0) {
      violations.push({
        code: "UNSOURCED_LINE",
        severity: "error",
        location,
        detail: "Rewritten line cites no source in the uploaded resume.",
        offending: text,
      });
    }

    const unknown = sourceIds.filter((id) => !factIndex.has(id));
    for (const id of unknown) {
      violations.push({
        code: "UNKNOWN_SOURCE_ID",
        severity: "error",
        location,
        detail: `Cites '${id}', which does not exist in the extracted resume facts.`,
        offending: text,
      });
    }

    const sourceText = sourceIds
      .filter((id) => factIndex.has(id))
      .map((id) => factIndex.get(id)!)
      .join(" ");
    const sourceDigits = new Set(extractMetrics(sourceText).map(digitsOf).filter(Boolean));

    for (const metric of extractMetrics(text)) {
      const d = digitsOf(metric);
      if (!d) continue;
      if (!sourceDigits.has(d) && !corpusDigits.has(d)) {
        violations.push({
          code: "UNSOURCED_METRIC",
          severity: "error",
          location,
          detail: `The figure '${metric}' does not appear anywhere in the uploaded resume.`,
          offending: text,
        });
      }
    }

    for (const tech of techTermsIn(text, vocabulary)) {
      if (!corpusTech.has(tech)) {
        violations.push({
          code: "UNSOURCED_TECH",
          severity: "error",
          location,
          detail: `'${tech}' is presented as the candidate's experience but is absent from the uploaded resume.`,
          offending: text,
        });
      }
    }
  };

  checkLine("Summary", tailored.summary.text, tailored.summary.sourceIds);

  // A headline speaks for the candidate's current level and claimed stack, so
  // it may draw on the most recent role and the skills section — not, as it
  // once did, on every role ever held, which let it borrow a technology from a
  // job three years ago and call it a specialisation.
  const headlineSources = [
    "HEADLINE",
    facts.experience[0]?.id,
    ...facts.skills.map((s) => s.id),
  ].filter((id): id is string => Boolean(id));
  checkLine("Headline", tailored.headline, headlineSources);

  for (const group of tailored.skills) {
    checkLine(`Skills / ${group.category}`, group.items.join(", "), group.sourceIds);
  }

  for (const exp of tailored.experience) {
    const fact = facts.experience.find((e) => e.id === exp.sourceId);
    if (!fact) {
      violations.push({
        code: "UNKNOWN_SOURCE_ID",
        severity: "error",
        location: `Experience / ${exp.company}`,
        detail: `Experience block cites '${exp.sourceId}', which is not an extracted role.`,
        offending: `${exp.title} at ${exp.company}`,
      });
      continue;
    }
    const drift: string[] = [];
    if (normalize(exp.company) !== normalize(fact.company)) drift.push(`company '${fact.company}' -> '${exp.company}'`);
    if (normalize(exp.title) !== normalize(fact.title)) drift.push(`title '${fact.title}' -> '${exp.title}'`);
    if (normalize(exp.startDate) !== normalize(fact.startDate)) drift.push(`start '${fact.startDate}' -> '${exp.startDate}'`);
    if (normalize(exp.endDate) !== normalize(fact.endDate)) drift.push(`end '${fact.endDate}' -> '${exp.endDate}'`);
    if (drift.length) {
      violations.push({
        code: "ALTERED_EMPLOYER_FACT",
        severity: "error",
        location: `Experience / ${fact.company}`,
        detail: `Verifiable employment details were changed: ${drift.join("; ")}.`,
        offending: drift.join("; "),
      });
    }
    exp.bullets.forEach((b, i) =>
      checkLine(`Experience / ${fact.company} / bullet ${i + 1}`, b.text, b.sourceIds),
    );
  }

  for (const proj of tailored.projects) {
    proj.bullets.forEach((b, i) =>
      checkLine(`Project / ${proj.name} / bullet ${i + 1}`, b.text, b.sourceIds),
    );
  }

  const errorCount = violations.filter((v) => v.severity === "error").length;
  const warningCount = violations.length - errorCount;

  return { passed: errorCount === 0, errorCount, warningCount, violations };
}

File: backend/src/atsresume/truth/guard.py
"""The truth guard.

Any model asked to optimise a resume for a job description will drift into
adding the keywords the description asks for. That is what gets a candidate
caught in an interview, and asking the model nicely not to do it is not a
control. So truthfulness is enforced here, after generation, deterministically.

Every rewritten line cites the ids of the resume facts it derives from, and each
citation is checked against the original text. Four things get caught:

    UNSOURCED_LINE        a line with no traceable origin
    UNKNOWN_SOURCE_ID     a citation to a fact that does not exist
    UNSOURCED_METRIC      a figure that was not in the uploaded resume
    UNSOURCED_TECH        a technology the candidate never claimed
    ALTERED_EMPLOYER_FACT a changed company, title, or employment date

The last one matters most in practice: employment details are the first thing a
recruiter verifies, and a promoted title is the fastest way to lose an offer.
"""

from __future__ import annotations

import re

from ..models import (
    ResumeFacts,
    TailoredResume,
    TruthReport,
    TruthViolation,
    ViolationCode,
)
from .vocabulary import build_vocabulary, implied_by, normalise, terms_present


class _CaseInsensitiveIndex(dict):
    """Fact ids, matched without regard to case."""

    def __setitem__(self, key: str, value: str) -> None:
        super().__setitem__(key.upper(), value)

    def __getitem__(self, key: str) -> str:
        return super().__getitem__(key.upper())

    def __contains__(self, key: object) -> bool:
        return isinstance(key, str) and super().__contains__(key.upper())

    def get(self, key: str, default=None):  # type: ignore[override]
        return super().get(key.upper(), default)

# Percentages, multipliers, durations, throughput, money, and any bare number of
# two digits or more. Single bare digits are skipped on purpose: "3 microservices"
# is usually a restatement, and flagging it buries the real findings in noise.
_METRIC = re.compile(
    r"""
    (?:\d[\d,.]*\s?
       (?:%|percent|x\b|ms\b|s\b|sec\b|seconds\b|min\b|minutes\b|hours?\b|days?\b
         |weeks?\b|months?\b|years?\b|k\b|m\b|b\b|mb\b|gb\b|tb\b|rps\b|qps\b|tps\b
         |req/s|users?\b|customers?\b|clients?\b))
  | (?:[$₹€£]\s?\d[\d,.]*)
  | (?:\b\d{2,}\b)
    """,
    re.IGNORECASE | re.VERBOSE,
)

_DIGITS = re.compile(r"[^0-9]")

# Lines that speak for the document rather than for one achievement, and may
# therefore draw a figure from anywhere in it.
_AGGREGATING_LOCATIONS = ("Summary", "Headline")


def _metrics(text: str) -> list[str]:
    return [m.group(0).strip().lower() for m in _METRIC.finditer(text)]


def _digits_of(value: str) -> str:
    return _DIGITS.sub("", value)


def build_fact_index(facts: ResumeFacts) -> dict[str, str]:
    """Every citable fact, by id.

    Keys are upper-cased on lookup because the model writes "summary" as often
    as "SUMMARY", and rejecting a correct citation over its capitalisation buys
    a repair round for nothing.
    """
    index: _CaseInsensitiveIndex = _CaseInsensitiveIndex()
    if facts.summary:
        index["SUMMARY"] = facts.summary
    if facts.headline:
        index["HEADLINE"] = facts.headline

    for exp in facts.experience:
        index[exp.id] = " ".join(
            [exp.company, exp.title, exp.location, exp.start_date, exp.end_date, *exp.tech]
        )
        for bullet in exp.bullets:
            index[bullet.id] = bullet.text
    for proj in facts.projects:
        index[proj.id] = " ".join([proj.name, proj.description, *proj.tech])
        for bullet in proj.bullets:
            index[bullet.id] = bullet.text
    for edu in facts.education:
        index[edu.id] = " ".join([edu.institution, edu.degree, edu.dates, edu.details])
    for group in facts.skills:
        index[group.id] = " ".join([group.category, *group.items])
    for cert in facts.certifications:
        index[cert.id] = cert.text
    for section in facts.other_sections:
        index[section.id] = section.heading
        for bullet in section.bullets:
            index[bullet.id] = bullet.text
    return index


def run_truth_guard(
    tailored: TailoredResume,
    facts: ResumeFacts,
    raw_resume_text: str,
    jd_terms: list[str] | None = None,
) -> TruthReport:
    violations: list[TruthViolation] = []
    index = build_fact_index(facts)
    vocabulary = build_vocabulary(jd_terms)

    corpus = normalise(" \n ".join(index.values()) + " \n " + raw_resume_text)
    corpus_tech = terms_present(corpus, vocabulary)
    # Owning a tool is claiming what it does, so the corpus covers the
    # capabilities its technologies imply as well as the names themselves.
    corpus_tech |= implied_by(corpus_tech)
    corpus_digits = {d for d in (_digits_of(m) for m in _metrics(corpus)) if d}

    # Which bullet has already claimed a given figure, so the same number cannot
    # be spread across a page as if it were several achievements.
    claimed_by: dict[str, str] = {}

    def check_line(location: str, text: str, source_ids: list[str]) -> None:
        if not text.strip():
            return

        if not source_ids:
            violations.append(
                TruthViolation(
                    code=ViolationCode.UNSOURCED_LINE,
                    location=location,
                    detail="Rewritten line cites no source in the uploaded resume.",
                    offending=text,
                )
            )

        for source_id in source_ids:
            if source_id not in index:
                violations.append(
                    TruthViolation(
                        code=ViolationCode.UNKNOWN_SOURCE_ID,
                        location=location,
                        detail=f"Cites '{source_id}', which is not an extracted resume fact.",
                        offending=text,
                    )
                )

        source_text = " ".join(index[s] for s in source_ids if s in index)
        source_digits = {d for d in (_digits_of(m) for m in _metrics(source_text)) if d}

        # A summary speaks for the whole document, so it may restate a figure
        # from any part of it. A bullet speaks only for its own achievement, and
        # must find its figures in the facts it actually cites. Accepting the
        # corpus everywhere was the guard's largest hole: a real 45% earned by
        # one piece of work licensed a fabricated 45% on a different one.
        aggregates = location.startswith(_AGGREGATING_LOCATIONS)
        is_bullet = " / bullet " in location

        for metric in _metrics(text):
            digits = _digits_of(metric)
            if not digits:
                continue

            if digits not in source_digits and not (aggregates and digits in corpus_digits):
                elsewhere = digits in corpus_digits
                violations.append(
                    TruthViolation(
                        code=ViolationCode.UNSOURCED_METRIC,
                        location=location,
                        detail=(
                            f"The figure '{metric}' appears elsewhere in the resume but not in "
                            f"the facts this line cites ({', '.join(source_ids) or 'none'})."
                            if elsewhere
                            else f"The figure '{metric}' does not appear anywhere in the "
                            "uploaded resume."
                        ),
                        offending=text,
                    )
                )
                continue

            # Uniqueness is a rule about bullets competing with one another. A
            # summary restating a bullet's number is how a good resume reads.
            if is_bullet:
                first = claimed_by.get(digits)
                if first is not None and first != location:
                    violations.append(
                        TruthViolation(
                            code=ViolationCode.DUPLICATED_METRIC,
                            location=location,
                            detail=(
                                f"The figure '{metric}' is already claimed at {first}. One "
                                "achievement stretched across two bullets reads as padding."
                            ),
                            offending=text,
                        )
                    )
                else:
                    claimed_by.setdefault(digits, location)

        for tech in terms_present(text, vocabulary):
            if tech not in corpus_tech:
                violations.append(
                    TruthViolation(
                        code=ViolationCode.UNSOURCED_TECH,
                        location=location,
                        detail=(
                            f"'{tech}' is presented as the candidate's experience but is absent "
                            "from the uploaded resume."
                        ),
                        offending=text,
                    )
                )

    # A headline speaks for the candidate's current level and claimed stack, so
    # it may draw on the most recent role and the skills section — not on every
    # role ever held, which would let it borrow a technology from years ago and
    # call it a specialisation.
    headline_sources = ["HEADLINE"]
    if facts.experience:
        headline_sources.append(facts.experience[0].id)
    headline_sources.extend(group.id for group in facts.skills)
    check_line("Headline", tailored.headline, [s for s in headline_sources if s in index])

    for location, text, source_ids in tailored.all_lines():
        check_line(location, text, source_ids)

    # Verifiable employment details must survive the rewrite untouched.
    by_id = {exp.id.upper(): exp for exp in facts.experience}
    for block in tailored.experience:
        fact = by_id.get(block.source_id.upper())
        if fact is None:
            violations.append(
                TruthViolation(
                    code=ViolationCode.UNKNOWN_SOURCE_ID,
                    location=f"Experience / {block.company}",
                    detail=f"Cites '{block.source_id}', which is not an extracted role.",
                    offending=f"{block.title} at {block.company}",
                )
            )
            continue

        drift: list[str] = []
        for label, was, now in (
            ("company", fact.company, block.company),
            ("title", fact.title, block.title),
            ("start", fact.start_date, block.start_date),
            ("end", fact.end_date, block.end_date),
        ):
            if normalise(was) != normalise(now):
                drift.append(f"{label} '{was}' -> '{now}'")
        if drift:
            violations.append(
                TruthViolation(
                    code=ViolationCode.ALTERED_EMPLOYER_FACT,
                    location=f"Experience / {fact.company}",
                    detail="Verifiable employment details were changed: " + "; ".join(drift) + ".",
                    offending="; ".join(drift),
                )
            )

    errors = sum(1 for v in violations if v.severity == "error")
    return TruthReport(
        passed=errors == 0,
        error_count=errors,
        warning_count=len(violations) - errors,
        violations=violations,
    )
